Created new API Tools page

mtesauro opened the pull request:
Since API tools are similar but not the same as web tools, I created a page to hold a list of API tools most of which I've been trying out so why not share the list more broadly than a text file on my computer.

This is a start on this page - there are definitely more tools than I have listed here. I plan on adding them as I come across them or try them out.  Hopefully the broader OWASP community will add to this list as well.

kingthorin commented:
@mtesauro thanks for this. Not sure how easily maintainable it'll be. But it can always be changed in the future. I'll test it tomorrow and hopefully get it merged.

mtesauro commented:
It's really easy to maintain in neovim ;-)

I debated easy to create vs looks nice and went for looks nice.  Happy to make adjustments and definitely wanting feedback.  Once the initial page gets created/finalized, I'll start adding more rows.

kingthorin commented:
Sadly your "Name" column isn't panning out (the line breaks aren't breaking and the links aren't linking):
![image](https://user-images.githubusercontent.com/7570458/150784688-00d26741-afd7-4823-9168-ccb36b31f85f.png)

I'd suggest switching to data file and an HTML include for the table, similar to:
- https://github.com/OWASP/www-community/blob/master/_data/tools.json
- https://github.com/OWASP/www-community/blob/master/_includes/tools.html
    - https://github.com/OWASP/www-community/blob/6434e614dc60c055e7b818da240bade93e70568a/pages/Vulnerability_Scanning_Tools.md
    -  https://github.com/OWASP/www-community/blob/6d6d9cfa162de634fea3b30df4a39d6209a744e5/pages/Source_Code_Analysis_Tools.md

Also you can find local testing instructions here: https://owasp.org/migration/#setup-a-local-testing-environment

kingthorin commented on pages/API_Security_Tools.md:
```diff
@@ -0,0 +1,66 @@
+author: Matt Tesauro
+contributors:
+tags: api, tools, oss
+permalink: /API_Security_Tools
```
Technically this works fine. However for user convenience I'd suggest using all lower instead of mixed case. (Same for the actual file name.)

mtesauro replied:
ToLower()'ed the permalink and file name.

mtesauro commented:
Great feedback.  Thanks!

Very strange about the links in the name column, I was :100: 'inspired' by [Steve's page](https://owasp.org/www-community/Component_Analysis) where it appears to be working.

Let me do some refactoring and I'll ping you here when that's done.  I am now tempted by Harold's not tables, table ;-)

Iterating through data files would definitely make updates easier.  I'll also setup the local testing stuff as I now know that what I get isn't how the page renders in my fork (how I was QA'ing this previously) cause this looks pretty in my fork:
![image](https://user-images.githubusercontent.com/3422419/150809514-a1f642d1-48b1-4b94-959a-bf863c89ad27.png)

kingthorin commented:
Ya how it renders in the GitHub web UI isn't necessarily how it'll render when deployed unfortunately.

Another option table wise is to go with `full-width` layout then there's no right nav bar taking up real estate.

kingthorin commented:
Also note the data file can also be yaml if you prefer. I went with json for the existing tools data file because it can then have a [job](https://github.com/OWASP/www-community/blob/master/.github/workflows/validate-tools-json.yml) to do schema validation pretty simply. [The job runs on PRs that change the file and validates that they conform to the schema, then comments on the PR if they don't.]

kingthorin commented on pages/api-security-tools.md:
```diff
@@ -0,0 +1,56 @@
+
+## API Tools List
+
+<section id="staff" class="corporate">
```
Could use an update vs copy/paste

kingthorin replied:
Throughout the table it seems, it’s not member data/classes

mtesauro replied:
TBH, this was an intentional cheat - I was attempting to use the CSS that's part of Staff pages.  Mostly, it's because I'm quite terrible at front-end/CSS work so, since leaving them there didn't seem to hurt, I never tried it without them.

Just tried removing the ids / classes and it doesn't appear to degrade how the page looks so I've removed them.

kingthorin commented on pages/api-security-tools.md:
```diff
@@ -0,0 +1,56 @@
+<br/><br/>
+</section>
+
+#### Adding tools
```
```suggestion
#### Adding Tools
```

mtesauro replied:
Done.

kingthorin commented on pages/api-security-tools.md:
```diff
@@ -0,0 +1,56 @@
+#### Adding tools
+
+To add items, please add a stanza to the yaml file [here](https://github.com/OWASP/www-community/blob/master/_data/api-tools.yml) or email me at matt.tesauro AT owasp.org
+<br/><br/>
```
```suggestion

```

mtesauro replied:
Replace with a div 'cheat' - my goal was to get a bit more separation between the end of the content I created and the auto-footer.

kingthorin commented on pages/api-security-tools.md:
```diff
@@ -0,0 +1,56 @@
+
+APIs are becoming an increasingly large portion of the software that powers the Internet including mobile applications, single-page applications (SPAs) and cloud infrastructure. While APIs share much of the same security controls and software security issues with traditional web applications, they are different enough to make a distinction between 'normal' AppSec tools and ones that were built wi ...
+
+### Types of API tools
```
```suggestion
### Types of API Tools
```

mtesauro replied:
Done.

mtesauro commented:
OK.  Week got away from me but I got a chance to finish this today.  @hblankenship @kingthorin 

Here's what's changed:

- Removed the table, switched to div's that are responsive so the data shows nicely in mobile and browsers
- Moved the former table data to a data yaml file `data/api-tools.yml` so updates only require adding bits to the yaml file
- Moved most of the data to sheilds.io 'icons' to make things more consistent
- Minor text changes to the non-data content including some advice on adding more tools.
- Created a Docker container to run Jekylly locally that does live reloads.  That made the zillion iterations on the 'look' of this page much nicer.  I've pushed the container to Docker Hub and made a public repo if anyone else wants to use it.  See https://github.com/mtesauro/owasp-test-site

Once this PR is merged, I'll work on adding more tools to the yaml file.  For reference, here's how it looks in a browser:
![image](https://user-images.githubusercontent.com/3422419/151716951-d85b3d86-3f7f-49a1-9948-dd4937209c8f.png)

And here's how it looks in mobile:
![image](https://user-images.githubusercontent.com/3422419/151716989-02cc9cb6-4647-4949-a5d1-96d2101fc62c.png)

All the 'fun' I had getting this to look nice reminded me why I usually create command-line tools ;-)

As always, feedback is always appreciated.

kingthorin commented:
Thanks @mtesauro few minor things.

Nice work!

mtesauro commented:
@kingthorin I suspect you get lots of people attempting to try out updates locally - I made a container for that makes it rather easy. The repo included running and building instructions. Feel free to share:

> Created a Docker container to run Jekylly locally that does live reloads. That made the zillion iterations on the 'look' of this page much nicer. I've pushed the container to Docker Hub and made a public repo if anyone else wants to use it. See https://github.com/mtesauro/owasp-test-site

kingthorin reviewed (approved):
Thanks!
